Dafny 3.4.2 Requires C# 8

<!-- This page is for Dafny bug reports and feature requests.  Please remember to include include your Dafny version number, any relevant code, and expected results. -->

<!-- Having trouble with a proof?  Consider posting your question on Stack Overflow. -->

While testing Dafny 3.4.2 on our library, which targets C#, we encountered the following error when testing in C#:

```
src/redacted/redacted/redacted/redacted/Source/obj/Debug/netstandard2.0/GeneratedFromDafny.cs(11926,20): error CS8370: Feature 'recursive patterns' is not available in C# 7.3. Please use language version 8.0 or greater.
```

There is no call out in the release notes for 3.4.2 that C#8 will now be required. 

Our library works fine on Dafny 3.4.1. 

Thank you!

--- a/RELEASE_NOTES.md
+++ b/RELEASE_NOTES.md
@@ -4,6 +4,7 @@
 - feat: The keyword syntax for functions will change in Dafny version 4. The new command-line option `/functionSyntax` (see `/help`) allows early adoption of the new syntax. (https://github.com/dafny-lang/dafny/pull/1832)
 - fix: No warning "File contains no code" if a file only contains a submodule (https://github.com/dafny-lang/dafny/pull/1840)
 - fix: export-reveals of function-by-method now allows the function body to be ghost (https://github.com/dafny-lang/dafny/pull/1855)
+- fix: Dafnyruntime's c# language version explicitly set to 7.1 to stay backward compatible. (https://github.com/dafny-lang/dafny/pull/1877)
 
 
 # 3.4.2

--- a/RELEASE_NOTES.md
+++ b/RELEASE_NOTES.md
@@ -4,7 +4,7 @@
 - feat: The keyword syntax for functions will change in Dafny version 4. The new command-line option `/functionSyntax` (see `/help`) allows early adoption of the new syntax. (https://github.com/dafny-lang/dafny/pull/1832)
 - fix: No warning "File contains no code" if a file only contains a submodule (https://github.com/dafny-lang/dafny/pull/1840)
 - fix: export-reveals of function-by-method now allows the function body to be ghost (https://github.com/dafny-lang/dafny/pull/1855)
-- fix: Regain C# 7.1 compatibility of the compiled code. (https://github.com/dafny-lang/dafny/pull/1877)
+- fix: Regain C# 7.3 compatibility of the compiled code. (https://github.com/dafny-lang/dafny/pull/1877)
 
 
 # 3.4.2

--- a/RELEASE_NOTES.md
+++ b/RELEASE_NOTES.md
@@ -4,7 +4,7 @@
 - feat: The keyword syntax for functions will change in Dafny version 4. The new command-line option `/functionSyntax` (see `/help`) allows early adoption of the new syntax. (https://github.com/dafny-lang/dafny/pull/1832)
 - fix: No warning "File contains no code" if a file only contains a submodule (https://github.com/dafny-lang/dafny/pull/1840)
 - fix: export-reveals of function-by-method now allows the function body to be ghost (https://github.com/dafny-lang/dafny/pull/1855)
-- fix: Dafnyruntime's c# language version explicitly set to 7.1 to stay backward compatible. (https://github.com/dafny-lang/dafny/pull/1877)
+- fix: Regain C# 7.1 compatibility of the compiled code. (https://github.com/dafny-lang/dafny/pull/1877)
 
 
 # 3.4.2

--- a/Source/DafnyRuntime/DafnyRuntime.cs
+++ b/Source/DafnyRuntime/DafnyRuntime.cs
@@ -1199,7 +1199,7 @@ namespace Dafny {
 
       while (toVisit.Count != 0) {
         var seq = toVisit.Pop();
-        if (seq is ConcatSequence<T> { elmts: { IsDefault: true } } cs) {
+        if (seq is ConcatSequence<T> cs && cs.elmts.IsDefault) {
           (leftBuffer, rightBuffer) = (cs.left, cs.right);
           if (cs.left == null || cs.right == null) {
             // !cs.elmts.IsDefault, due to concurrent enumeration

--- a/Source/DafnyRuntime/DafnyRuntime.csproj
+++ b/Source/DafnyRuntime/DafnyRuntime.csproj
@@ -9,7 +9,7 @@
       <PackageVersion>1.1.0</PackageVersion>
       <TargetFramework>net6.0</TargetFramework>
       <OutputPath>..\..\Binaries\</OutputPath>
-      <LangVersion>7.1</LangVersion>
+      <LangVersion>7.3</LangVersion>
   </PropertyGroup>
 
   <ItemGroup>

--- a/Source/DafnyRuntime/DafnyRuntime.csproj
+++ b/Source/DafnyRuntime/DafnyRuntime.csproj
@@ -9,6 +9,7 @@
       <PackageVersion>1.1.0</PackageVersion>
       <TargetFramework>net6.0</TargetFramework>
       <OutputPath>..\..\Binaries\</OutputPath>
+      <LangVersion>7.1</LangVersion>
   </PropertyGroup>
 
   <ItemGroup>

